### DISCUSSION TITLE
How to perform complete LazyVim setup from build script?

### DISCUSSION THREAD

**Discussion by pwnalone** [LazyVim/LazyVim#3679] [Q&A]
I'm trying to create a containerized dev environment for myself that comes with Neovim already setup with all Neovim plugins, LSP servers, and Treesitter parsers installed. So far, the best I've got is executing the following from a build script that is run from my container's `Dockerfile`, which installs all my Neovim plugins, but does not install the LSP servers or Treesitter parsers.

```sh
nvim --headless '+Lazy install' +q
```

I tried adding `'+TSInstallSync all'` to the command line before the `+q`, but that didn't work, and I'm not even sure how you would get Mason to install all configured LSP servers (plus linters / formatters) without adding a `'+MasonInstall {tool}'` for each tool.

Any help getting this working would be much appreciated.

**Comment by dpetka2001** [comment]
Never done this, but found a related discussion [here](https://github.com/folke/lazy.nvim/discussions/1188). May be of help to what you want to do.

**Reply by pwnalone** [reply]
This isn't really what I want because I'd have to update my `Dockerfile` every time I add a new Mason tool to be installed by my Neovim config. However, that link did lead me to this Github [issue](https://github.com/williamboman/mason.nvim/issues/467), which showed me how to create a user command to install all tools listed in the `ensure_installed` table.

With the `lazy` package manager, however, this still isn't ideal because multiple plugin configs will update this table when loaded, so I'd need to somehow load all plugins that might update this table before invoking my user command to ensure that all tools get installed. So, for now, I'm opting to repeat the list of Mason tools that I would want to install in the Mason portion of my Neovim config.

```lua
local Util = require("util.common")

return {
  {
    "williamboman/mason.nvim",
    cmd = { "MasonInstallAll" },
    opts = function(_, opts)
      local ensure_installed = {
        "clangd",
        "codelldb",
        "debugpy",
        "lua-language-server",
        "pyright",
        "ruff-lsp",
        "rust-analyzer",
        "shfmt",
        "stylua",
        "taplo",
      }

      opts.ensure_installed = Util.tbl_dedup(vim.tbl_extend("force", opts.ensure_installed, ensure_installed))

      -- Create user command to synchronously install all Mason tools in `opts.ensure_installed`.
      vim.api.nvim_create_user_command("MasonInstallAll", function()
        for _, tool in ipairs(opts.ensure_installed) do
          vim.cmd("MasonInstall " .. tool)
        end
      end, {})

      return opts
    end,
  },
}
```

And then I can install all my plugins, LSP/DAP servers, linters, and formatters like so:

```sh
nvim --headless '+Lazy install' +MasonInstallAll +qall
```

**Comment by kureta** [comment]
I have been slowly going insane in the past few days. Mainly because I don't know any lua and had to learn some, and just wing it for the rest. I couldn't find a way to sleep/wait in a lua script without blocking the entire neovim, which is useless. I also couldn't find a way to listen for an event or something to run a function, for example, after Lazy plugins are loaded, after mason is ready, after all mason install operations are completed, etc.

So my solution is a bit hacky, probably depends on some race conditions, but maybe not. I looked at the internals of `lazy.nvim` found a way to access the currently loaded final options, gathered from every config file that is loaded. I don't want to manually list all mason stuff in my config. I just add `{ import = "lazyvim.plugins.extras.lang.python" }` to my plugins and it installs pyright, ruff, and other stuff. No additional setup should be necessary. Anyway, as far as I can tell, there are two places where these configurations that automatically install mason stuff end up: `nvim-lspconfig` and `mason`.

```lua
local lazy = require("lazy.core.config")
local nvim_lsp = lazy.plugins["nvim-lspconfig"]
local mason = lazy.plugins["mason.nvim"]
local registry = require("mason-registry")

-- lsp packages installed via nvim-lspconfig
local servers = nvim_lsp._.cache.opts.servers

-- mason apps installed
local mason_apps = mason._.cache.opts.ensure_installed
```

This way I can access a complete list of mason stuff that are configured to be autoinstalled. But for some reason, some packages have "aliases" and these lists sometimes contain aliases instead of actual package names. So I had to create a table to convert those to package names:

```lua
-- Stuff below is for converting package aliases to names
local registry = require("mason-registry")

local all_package_names = registry.get_all_package_names()
local alias_to_name = {}

for _, package_name in pairs(all_package_names) do
	local aliases = registry.get_package_aliases(package_name)
	if aliases then
		for _, alias in pairs(aliases) do
			alias_to_name[alias] = package_name
		end
		alias_to_name[package_name] = package_name
	end
end
```

Then, finally I can install all of them. ~~The last problem, however is that I cannot wait for any event to complete. If  add `vim.cmd("qa")` at the end of the script it just quits while mason is still installing stuff.~~

Found a way to block until `MasonInstall` completes and put `Lazy! install all` into the `install.lua` script. No more hacks needed. [Here is everything](https://github.com/kureta/dev) required to build a docker image with this. I was really furstrated and in a hurry, so I might have pushed some secrets or personal information to the repo. If you see anything like that please let me know.

**Reply by ryanpeach** [reply]
Great start IMO. Doesn't work for me personally but I cant seem to get it to work with "nvim --headless '+Lazy install' +qall" in general yet.

**Reply by kureta** [reply]
Did you look at the complete example in [my repo](https://github.com/kureta/dev)? What is not working? If you give more info, maybe I can help.

**Comment by ryanpeach** [comment]
I’ll make a PR to my repo after reviewing yours and send it to you. May be a while though. Thanks for the help!

**Comment by lukas-zenner-webcom** [comment]
Not sure if this is still a problem.
I wanted to have lazy vim inside a docker container.

I ended up with creating a docker volume and mounting it to `/root/.local`.
That way I persisted the installation results.
